docs/gadget-devel/metrics.md

```diff
@@ -0,0 +1,144 @@
+choose depends on how you collect the data and the quantity of it.
+
+If you're sending events to user-space, you can create metrics from those by just adding a couple of annotations to your
+`gadget.yaml` file, or by using some [well-known types](!LINK_TO_MACROS_AND_TYPES) inside your struct definition inside eBPF code.
+
+If you don't want to emit events, because it would just be too much throughput, you can choose to write the metrics into
+eBPF maps instead and let IG create a [data source](!LINK_TO_DATASOURCE_EXPLANATION) from it. Those can then be
+exported to for example Prometheus.
+
+| Source     | Application                                                                             | Performance |
+|------------|-----------------------------------------------------------------------------------------|-------------|
+| user-space | easy development when using event sources (GADGET_TRACE()); event sources can be reused | slower      |
+| eBPF maps  | using dedicated maps for metrics, especially suited for high throughput scenarios       | fast        |
+
+## User-space collection
+
+### Without eBPF
+
+Without modifying eBPF code, you can modify the `gadget.yaml` to specify how metrics should be collected from existing
+data sources. Keep in mind that this requires an event being sent to user-space in order to work - so if you expect
+many events, please look into implementing the metrics collection using eBPF maps (see below).
+
+Let's assume you've already got a data source called `events`. That data source has two fields - a string field called
+`name` and an int32 field called `count`. The (empty) section of your `gadget.yaml` file for that datasource could look
+like this:
+
+```yaml
+datasources:
+  events:
+    fields:
+      name:
+      count:
+```
+
+First, you need to annotate the data source itself with `metrics.collect` set to `true`.
+
+Now, if you want to expose the field `count` as a metric, just annotate the field with `metrics.type` as key and `counter`
+as value, like so:
+
+```yaml
+datasources:
+  events:
+    annotations:
+      metrics.collect: true
+    fields:
+      count:
+        annotations:
+          metrics.type: counter
+```
+
+This will increase the counter of that metrics by whatever value the `count` field has. Use `metrics.type: gauge`, if
+you always just want to set the metric to the latest value.
+
+If you later on want to differentiate the count by their different `name` contents, you can annotate the `name` field
+with key `metrics.type` and value `key`. This creates a key (or label) from that field that is associated with each
+metric afterward - e.g.: this lets you later on filter the values of `count` according to the corresponding `name`
+values.
+
+> Note: adding multiple metric.key annotations will increase cardinality a lot and take up more space and storage.
+
+### With eBPF
```

Comment:
I got a bit confused by this title. In this case ebpf is only used to annotate the fields, but the collection is happening in user-space. I'd suggest to name these two like:
- Using the metadata file
- Using well-known types in the eBPF code

Reply:
Much better, thanks!